### ISSUE TITLE
UDP connection can get stuck and not reaching a service's pod if initiated during this pod restart

### ISSUE THREAD

**Issue by akantek-wr** [kubernetes/kubernetes#114668]
### What happened?

It was observed that on the following scenario: [pod-A] => [service-B] => [pod-B] 
  - If pod-B restarts usually pod-A starts a UDP flow at a rate of ~1pkt/s
  - Sometimes the connection get stuck without executing the service-B DNAT rule to reach pod-B

I dont have access to the real tenant's application, as an example I created this scenario (attached k8s yaml):
```
[root@controller-1 sysadmin(keystone_admin)]# kubectl get all -o wide
NAME                          READY   STATUS    RESTARTS   AGE    IP              NODE           NOMINATED NODE   READINESS GATES
pod/testpod1                  1/1     Running   0          4h2m   172.16.192.86   controller-0   <none>           <none>
pod/testpod2-5c44d866-jcb6k   1/1     Running   0          120m   172.16.192.93   controller-0   <none>           <none>
NAME                   TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE     SELECTOR
service/kubernetes     ClusterIP   10.96.0.1       <none>        443/TCP             5h56m   <none>
service/testpod2-udp   ClusterIP   10.108.141.11   <none>        5201/UDP,5201/TCP   4h2m    app=testpod2
NAME                       READY   UP-TO-DATE   AVAILABLE   AGE    CONTAINERS   IMAGES         SELECTOR
deployment.apps/testpod2   1/1     1            1           4h2m   testpod2     centos/tools   app=testpod2
NAME                                DESIRED   CURRENT   READY   AGE    CONTAINERS   IMAGES         SELECTOR
replicaset.apps/testpod2-5c44d866   1         1         1       4h2m   testpod2     centos/tools   app=testpod2,pod-template-hash=5c44d866 
```
Prior to the problem, pod-A did not have any UDP connections:
```
controller-0:/home/sysadmin# conntrack -L  -p udp -s 172.16.192.86 -d 10.108.141.11
conntrack v1.4.4 (conntrack-tools): 0 flow entries have been shown. 
```
And I started to monitor iptables events with:
`xtables-monitor -e | while IFS= read -r line; do printf '[%s] %s\n' "$(date '+%Y-%m-%d %H:%M:%S:%N')" "$line"; done
`

If the pod-B (testpod2-5c44d866-jcb6k) restarts the following events happen (filtered to the relevant rules, but full log attached):

- this first rule allows sending ICMP unreachable back to the sender since the pod-B will not be available:
`[2022-06-23 16:46:52:645996593]  EVENT: -4 -t filter -A KUBE-SERVICES -d 10.108.141.11/32 -p udp -m comment --comment "default/testpod2-udp:testpood2-udp has no endpoints" -m udp --dport 5201 -j REJECT --reject-with icmp-port-unreachable`

- Then calico removes the old NAT rule:
`[2022-06-23 16:46:52:796939314]  EVENT: -4 -t nat -D KUBE-SEP-CE25SXMEH3YUH2JT -p udp -m comment --comment "default/testpod2-udp:testpood2-udp" -m udp -j DNAT --to-destination 172.16.192.93:5201 
`
After this removal, any packet sent by pod-A will receive a port unreachable response, this usually closes the socket on the sender side, and if packets are sent they will not trigger an entry on the conntrack table.

The state of conntrack:
```
controller-0:/home/sysadmin# conntrack -L  -p udp -s 172.16.192.86 -d 10.108.141.11
conntrack v1.4.4 (conntrack-tools): 0 flow entries have been shown. 
```

- Then, with the new pod up it is time to remove the REJECT rule:
`[2022-06-23 16:46:56:351623829]  EVENT: -4 -t filter -D KUBE-SERVICES -d 10.108.141.11/32 -p udp -m comment --comment "default/testpod2-udp:testpood2-udp has no endpoints" -m udp --dport 5201 -j REJECT --reject-with icmp-port-unreachable` 

Here, for example, pod-A starts generating UDP traffic towards service-B/pod-B, and keeps sending at ~1 pkt/s. At this point a new connection will be created, but instead of having the reversed path with the new pod-B address, it has the host local address (10.20.2.4):
```
controller-0:/home/sysadmin# conntrack -L  -p udp -s 172.16.192.86 -d 10.108.141.11
udp      17 29 src=172.16.192.86 dst=10.108.141.11 sport=50001 dport=5201 [UNREPLIED] src=10.108.141.11 dst=10.20.2.4 sport=5201 dport=39821 mark=0 use=1
conntrack v1.4.4 (conntrack-tools): 1 flow entries have been shown. 
```
**This is expected as there is no DNAT rule yet** to direct the packets to pod-B

- Observe that the new nat rule is installed after ~268ms (plenty of time on an automated system):
`[2022-06-23 16:46:56:619937072]  EVENT: -4 -t nat -A KUBE-SEP-OIKMPWYTENZYCWSH -p udp -m comment --comment "default/testpod2-udp:testpood2-udp" -m udp -j DNAT --to-destination 172.16.192.94:5201 
`
Since pod-A keeps sending the packet conntrack never resets (the timer is always at 29s), the new DNAT rule does not affect this connection, would only to new ones.

```
controller-0:/home/sysadmin# conntrack -L  -p udp -s 172.16.192.86 -d 10.108.141.11
udp      17 29 src=172.16.192.86 dst=10.108.141.11 sport=50001 dport=5201 [UNREPLIED] src=10.108.141.11 dst=10.20.2.4 sport=5201 dport=39821 mark=0 use=1
conntrack v1.4.4 (conntrack-tools): 1 flow entries have been shown. 
```
Of course one way to end this condition would be to stop the traffic for 30sec to clean the old connection, but it is not desirable
[test_pods_udp.yaml.txt](https://github.com/kubernetes/kubernetes/files/10289915/test_pods_udp.yaml.txt)
[xtables-monitoe_events.txt](https://github.com/kubernetes/kubernetes/files/10289917/xtables-monitoe_events.txt)

### What did you expect to happen?

In my opinion, the REJECT rules should only be removed after the new DNAT rule is installed to close this potential gap.

### How can we reproduce it (as minimally and precisely as possible)?

Since the bug occurs intermittently it may require several attempts

1. You need just a single node system
2. bringup the scenario 
3. install scapy on testpod1 
```
   $ yum install -y wget
   $ wget --trust-server-names https://github.com/secdev/scapy/archive/master.zip   # or wget -O master.zip https://github.com/secdev/scapy/archive/master.zip
   $ unzip master.zip
   $ cd master
   $ cd scapy
   $ sudo python setup.py install
   $ enter scapy and prepare the command:
     >> sendp(Ether()/IP(src="[pod-A IP]",dst="[service-B IP]")/UDP(sport=50000,dport=5201)/DNS(), iface="eth0", inter=0.001, count=500000)
```
4. on testpod2 start a ncat server
`   $ ncat -l 5201 -u -4 --keep-open --exec "/bin/echo Hello from testpod2." 
`
5. on the host start monitoring iptables to get the moment REJECT is installed
`   $ xtables-monitor -e | while IFS= read -r line; do printf '[%s] %s\n' "$(date '+%Y-%m-%d %H:%M:%S:%N')" "$line"; done | grep ""default/testpod2-udp:testpood2-udp has no endpoints" -m udp --dport 5201 -j REJECT"
`
6. restart pod-B 
`   $ kubectl delete pod/testpod2-[hash value]
`
7. As soon xtables-monitor shows the REJECT rule instalation start the traffic with scapy on testpod1
`   >> sendp(Ether()/IP(src="[pod-A IP]",dst="[service-B IP]")/UDP(sport=50000,dport=5201)/DNS(), iface="eth0", inter=0.001, count=500000)
`
8. You should see this unwanted connection on conntrack

Alternativelly, since there is a time gap, you can manually execute the iptables rules just to verify that there is the possibility of an unwanted UDP connection from pod-A to the service-B be created on the gap
- add the REJECT rule
`iptables -4 -t filter -A KUBE-SERVICES -d [service-B IP]/32 -p udp -m comment --comment "default/testpod2-udp:testpood2-udp has no endpoints" -m udp --dport 5201 -j REJECT --reject-with icmp-port-unreachable
`
The state of conntrack:
```
controller-0:/home/sysadmin# conntrack -L  -p udp -s [pod-A IP] -d [service-B IP]
conntrack v1.4.4 (conntrack-tools): 0 flow entries have been shown. 
```
- Remove the DNAT rule
`iptables -4 -t nat -D KUBE-SEP-OIKMPWYTENZYCWSH -p udp -m comment --comment "default/testpod2-udp:testpood2-udp" -m udp -j DNAT --to-destination [pod-B IP]:5201 
`
Start the testpod1 traffic here, it can be a more relaxed rate:
`>> sendp(Ether()/IP(src="172.16.192.81",dst="10.106.131.165")/UDP(sport=50000,dport=5201)/DNS(), iface="eth0", inter=1, count=500)
`
The unwanted connection is created:
```
controller-0:/home/sysadmin# conntrack -L  -p udp -s [pod-A IP] -d [service-B IP]
udp      17 29 src=[pod-A IP] dst=[service-B IP] sport=50001 dport=5201 [UNREPLIED] src=[service-B IP] dst=[local host IP]] sport=5201 dport=39821 mark=0 use=1
conntrack v1.4.4 (conntrack-tools): 1 flow entries have been shown. 
```
- remove the REJECT rule:
`iptables -4 -t filter -D KUBE-SERVICES -d [service-B IP]/32 -p udp -m comment --comment "default/testpod2-udp:testpood2-udp has no endpoints" -m udp --dport 5201 -j REJECT --reject-with icmp-port-unreachable 
`
- entering the new DNAT rule:
`iptables -4 -t nat -A KUBE-SEP-OIKMPWYTENZYCWSH -p udp -m comment --comment "default/testpod2-udp:testpood2-udp" -m udp -j DNAT --to-destination [pod-B IP]:5201 
`
Will not create a new entry in conntrack, the old one will remain until the traffic is stopped to give it time to expire
```
controller-0:/home/sysadmin# conntrack -L  -p udp -s [pod-A IP] -d [service-B IP]
udp      17 29 src=[pod-A IP] dst=[service-B IP] sport=50001 dport=5201 [UNREPLIED] src=[service-B IP] dst=[local host IP]] sport=5201 dport=39821 mark=0 use=1
conntrack v1.4.4 (conntrack-tools): 1 flow entries have been shown. 

```

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.8", GitCommit:"4a3b558c52eb6995b3c5c1db5e54111bd0645a64", GitTreeState:"archive", BuildDate:"2022-07-28T15:35:09Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.8", GitCommit:"4a3b558c52eb6995b3c5c1db5e54111bd0645a64", GitTreeState:"clean", BuildDate:"2021-12-15T14:46:22Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
[starlingx](https://www.starlingx.io/)
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
controller-0:/home/sysadmin# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

$ uname -a
Linux controller-0 5.10.112-200.23.tis.el7.x86_64 #1 SMP PREEMPT Wed Nov 2 15:31:22 EDT 2022 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
All the UDP conntrack bugs in kube-proxy were fixed in the recent versions, `v1.21.8` is EOL, you should try with a more recent version

**Comment by fruitbox12** [comment]
It looks like you are experiencing an issue with the networking configuration in your Kubernetes cluster, where packets are not being correctly forwarded to the destination pod.

One possible cause of this issue is a problem with the Network Address Translation (NAT) rules in the cluster. NAT rules are used to map the source and destination addresses and ports of packets, and they are typically used to allow communication between pods in different network namespaces.

In your scenario, it appears that the NAT rule for the testpod2-udp service is being removed when the destination pod (testpod2-5c44d866-jcb6k) restarts. This can cause the packets from the source pod (testpod1) to be dropped, as they do not have a valid route to the destination.

To troubleshoot this issue, you can try the following steps:

Check the logs of the networking components in your cluster, such as the networking plugin and the kube-proxy component. These logs may contain information about the NAT rules being created and deleted, and they may provide clues about the cause of the issue.
Check the configuration of the networking plugin and the kube-proxy component to ensure that they are correctly configured and functioning as expected.
If you are using a third-party networking solution, such as Calico, check the documentation and logs of that solution to see if it has any known issues that could be causing this problem.
I hope these suggestions are helpful! Let me know if you have any questions or if you need further assistance.
